### PR TITLE
Remove component union types

### LINKED ISSUE #6871: Remove component `union` types
See: https://github.com/WebAssembly/component-model/issues/236

### PR DESCRIPTION
**Pull request by lann**
...and bump wasm-tools deps.

Fixes #6871

### CODE REVIEW

**Comment by lann** [comment]
Assuming it also gets merged, the first commit here should be refactored to use vhttps://github.com/bytecodealliance/wasm-tools/pull/1182

**Review by alexcrichton — APPROVED** [review]

File: cranelift/wasm/src/sections_translator.rs [hunk 1]
```diff
@@ -50,20 +49,9 @@ pub fn parse_type_section<'a>(
...
-                let ty = environ.convert_func_type(&wasm_func_ty);
-                environ.declare_type_func(ty)?;
-            }
-            StructuralType::Array(_) | StructuralType::Struct(_) => {
-                unimplemented!("gc proposal");
-            }
-        }
+    for ty in types.into_iter_err_on_gc_types() {
+        let ty = environ.convert_func_type(&ty?);
+        environ.declare_type_func(ty)?;
```
Comment (lann) [inline cranelift/wasm/src/sections_translator.rs:54]: @alexcrichton Updated this

**Review by alexcrichton — APPROVED** [review]

File: crates/environ/src/module_environ.rs [hunk 2]
```diff
@@ -237,20 +237,9 @@ impl<'a, 'data> ModuleEnvironment<'a, 'data> {
...
-                            let ty = self.convert_func_type(&wasm_func_ty);
-                            self.declare_type_func(ty)?;
-                        }
-                        StructuralType::Array(_) | StructuralType::Struct(_) => {
-                            unimplemented!("gc proposal")
-                        }
-                    }
+                for ty in types.into_iter_err_on_gc_types() {
+                    let ty = self.convert_func_type(&ty?);
+                    self.declare_type_func(ty)?;
```
Comment (lann) [inline crates/environ/src/module_environ.rs:242]: ...and this

**Comment by lann** [comment]
I just pushed an update here anticipating a wit-bindgen 0.11.0 release. Obviously CI failed, but someone with access can hopefully just retrigger the CI jobs after that release and this should then be mergeable.

**Comment by alexcrichton** [comment]
At this rate I think I'll post a bump for 0.11.0 on Monday, but it'll definitely include this

**Comment by alexcrichton** [comment]
Oh oops wrong project, I see now! I think it's ok to avoid a wit-bindgen update here as that's mainly used for tests. The dependency between wit-bindgen and wasmtime mainly goes the other direction where wit-bindgen relies on wasmtime for tests, but it's ok in this regard for them to be slightly out of sync.

**Comment by lann** [comment]
Ah yeah I didn't really think that through. Well I'll fix it up when I get a chance if someone else doesnt get to it.

**Comment by lann** [comment]
This should be ready now.

**Review by alexcrichton — APPROVED** [review]

**Pull request merged**